pkg/kubelet/cpumanager/cpu_manager.go

```diff
@@ -0,0 +1,363 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpumanager
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/record"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
+	"k8s.io/kubernetes/pkg/kubelet/cpumanager/state"
+	"k8s.io/kubernetes/pkg/kubelet/cpumanager/topology"
+	"k8s.io/kubernetes/pkg/kubelet/eviction"
+	"k8s.io/kubernetes/pkg/kubelet/lifecycle"
+	"k8s.io/kubernetes/pkg/kubelet/status"
+	"k8s.io/kubernetes/pkg/kubelet/util/format"
+)
+
+type kletGetter interface {
+	GetPods() []*v1.Pod
+	GetCachedMachineInfo() (*cadvisorapi.MachineInfo, error)
+}
+
+type runtimeService interface {
+	UpdateContainerResources(id string, resources *runtimeapi.LinuxContainerResources) error
+}
+
+type policyName string
+
+// Manager interface provides methods for kubelet to manage pod cpus
+type Manager interface {
+	Start()
+
+	// RegisterContainer registers a container with the cpuset manager
```

Comment:
Why not have a registration phase at the pod creation time?
For example, kubelet can invoke podContainerManager->EnsureExists(pod) which can ensure that in addition to creating and setting up pod level cgroup, CPU allocations are feasible and even record CPU allocations early on.
If CPU allocations aren't possible (cross socket for example), then the pod can be rejected (this isn't possible yet since EnsureExists is invoked in the sync loop). 
If CPU allocation succeeds, then the podContainerManager can trigger rearrangement of CPU sockets across Bu and BE pods in the background while bringing the node to the desired state. 
This way there isn't a failure possible during container startup. And container restarts do not have to re-register themselves. 
This model will also pave way for identifying devices that are on the same socket as CPUs. For example, get all possible CPU and device allocations and choose the most appropriate combination, then go ahead and reserve CPU and devices.